### PR TITLE
Cargo Update: Cargo Supplypod Beacons! (THE CODEX ASTARTES CALLS THIS MANEUVER: STEEL RAIN)[New sprite!]

### PR DESCRIPTION
**Pull request by MrDoomBringer**
## [video of it in action!](https://youtu.be/_bM_Dw8QENM)
old_sprite
image

**why:**

-  Adds more game progression to cargo, giving them the ability to print beacons which can be used to direct supplypods to. 
- Also encourages teamwork within cargo, encouraging cargonians to do their job and talk to each other rather than fucking off and not communicating, etc.

**Current Balancy things in place:**
1. Beacons are pretty weak, and will be destroyed if they get hit by ~two pods
2. Beacons do not work when inside of any container (except for people's hands)
3. Beacons require two people to use effectively (one at the console and one holding/placing the beacon)
4. Express consoles can only have one beacon connected at a time. 
5. Beacons cost 5000 credits
6. Bluespace pods now come in 1.5 seconds faster

:cl: MrDoomBringer
add: Cargo now has access to Supplypod Beacons! Use these to call down supplypods with frightening accuracy!
/:cl:

### **the people:** _requests features_
image
image
image
image

image
### **me:** _adds the features_

anyways, im interested in hearing your opinions on this! Is it good where it stands? should x or y be added?

_~note: @WJohn said he doesnt like the beacon sprites, so i'll be making new ones. Consider the current ones to be a placeholder~ he likes the new sprites!
note 2: i added a funeral pack so you can quickly bury those crushed underneath supplypods
note 3: i also made bluespace pods come down 1.5 seconds faster
note 4: coffins were refactored from closets to crates. They sell for 250 credits._

### CODE REVIEW

**Comment by Metroidlover** [comment]
Wow, that was fast

**Comment by MrDoomBringer** [comment]
@arsserpentarium that can be solved by upgrading the console with techwebs!

**Comment by arsserpentarium** [comment]
Oh.This way it's fine.

**Comment by GunHog** [comment]
Drop a pod full of medical supplies directly onto the CMO! **Aggressive marketing!**

**Comment by Bombpants** [comment]
I have a question about how the pods work, does sending fuel canisters down have them detonate upon impact?

**Comment by kevinz000** [comment]
:+1: good pr by a very good player and QM.

**Comment by Poojawa** [comment]
This should be able to be placed in circuits. so you can give someone a meme 'flashlight' and then murder them with cargo pods.

**Comment by Shapsy** [comment]
oh my god you weren't kidding with the steel rain thing

**Comment by MrDoomBringer** [comment]
@kevinz000 aaaaaa

**Comment by praisenarsie** [comment]
now fix crateconomy

**Comment by praisenarsie** [comment]
also why doesn't the express pod gib you or instakill you, i expect to be fucking crushed not barely survive.

same goes for the beacon, why isn't that crushed too.

File: code/modules/cargo/expressconsole.dm [hunk 1]
```diff
@@ -1,31 +1,46 @@
...
 	blockade_warning = "Bluespace instability detected. Delivery impossible."
 	req_access = list(ACCESS_QM)
 	var/message
-	var/locked = TRUE
+	var/printed_beacons = 0 //number of beacons printed. Used to determine beacon names.
 	var/list/meme_pack_data
-	var/podID = 0//0 is your standard supply droppod (requires dissassembly after landing), 1 is the bluespace drop pod (teleports out after landing)
+	var/obj/item/device/supplypod_beacon/beacon //the linked supplypod beacon
+	var/area/landingzone = /area/quartermaster/storage //where we droppin boys
+	var/podID = 0 //0 is your standard supply supplypod (requires dissassembly after landing), 1 is the bluespace supply pod (teleports out after landing)
```
Comment (ShizCalev) [inline code/modules/cargo/expressconsole.dm]: make this a define

File: code/modules/cargo/expressconsole.dm [hunk 2]
```diff
@@ -86,11 +116,28 @@
 		packin_up()
 		stack_trace("You didn't give the cargo tech good advice, and he ripped the manifest. As a result, there was no pack data for [src]")
 	data["supplies"] = meme_pack_data
-
+	if (cooldown > 0)
+		cooldown--
 	return data
 
 /obj/machinery/computer/cargo/express/ui_act(action, params, datum/tgui/ui)
 	switch(action)
```
Comment (ShizCalev) [inline code/modules/cargo/expressconsole.dm]: does this fail is the mob is dead? don't see any canUseTopic checks anywhere

File: code/modules/cargo/supplypod_beacon.dm [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+
+/obj/item/device/supplypod_beacon/afterattack(obj/O, mob/user, proximity)
+	if(!istype(O) || !proximity)
+		return
+	if(istype(O, /obj/machinery/computer/cargo/express))
+		if (express_console != O)
+			link_console(O)
+
+/obj/item/device/supplypod_beacon/AltClick(mob/user)
+	if (express_console)
```
Comment (ShizCalev) [inline code/modules/cargo/supplypod_beacon.dm]: need a canUseTopic check before this, otherwise ghosts/dead mob/ect can do it from anywhere.
canUseTopic(src, !issilicon(user))

**Review by ShizCalev — CHANGES_REQUESTED** [review]

**Comment by Kromgar** [comment]
Add a suicide ti it where a pod lands and gibs you. The pod contains nothing.

**Comment by MrDoomBringer** [comment]
@Kromgar I'll add that in a separate PR along with an admin smite for it

File: code/modules/cargo/expressconsole.dm (hunk 2 above)
Comment (MrDoomBringer) [reply code/modules/cargo/expressconsole.dm]: @ShizCalev yes it does, because its a subtype of console. (just tested it)

**Comment by Vivalas** [comment]
Does this contain the thing where you can drop anywhere by emagging express console? I feel like nuking the station with fuel pods will get old after a while.

**Comment by MrDoomBringer** [comment]
@Vivalas no, I'd rather that be in a seperate PR.

**Comment by nicbn** [comment]
I'm against having it to drop from the sky. I'd rather see some kind of teleporter being used.

**Comment by MrDoomBringer** [comment]
@nicbn supplypods are already in game, though. This PR doesnt change or modify that in any way. Also, there are a few balance things that stem from it falling from the sky and creating an explosion when it lands

**Comment by AlexTheSysop42** [comment]
What happens when you order a pod:
1. Centcom loads it
2. They load it into a railgun aimed at your station
3. The pod is fired at many times light speed to reach your station in seconds
4. The pod slows down using some fancy tech or something 
5. The pod crashes through your station's roof and lands
6. Pod opens
6a. If it's a bluespace pod it teleports away

**Comment by arsserpentarium** [comment]
Then it need to depressurize section.

**Comment by AlexTheSysop42** [comment]
5k credits as well as 2 uses each (unless you can move them to save them?) seems a bit pricey

**Comment by AlexTheSysop42** [comment]
@arsseptenarium the station has a self-reparing roof, why do you think you only ever need to repair the floor?

That tech doesn't work on the floor though because pipes

**Comment by MrDoomBringer** [comment]
@AlexTheSysop42 You can move them to save them! But you gotta be quick on ya feet.

**Comment by AlexTheSysop42** [comment]
Or you can just slap them on a conveyor or disposals loop

**Comment by MrDoomBringer** [comment]
@AlexTheSysop42 Conveyor yes, but a disposals loop will not work, as the beacon must be on a turf, not in a container of any sort.

**Comment by optimumtact** [comment]
conflict

**Comment by MrDoomBringer** [comment]
@optimumtact i cast conflictus deletus

**Comment by MrDoomBringer** [comment]
@mapdiffbot is in need of help i think

File: code/game/objects/structures/crates_lockers/closets/job_closets.dm [hunk 4]
```diff
@@ -125,17 +125,6 @@
 	new /obj/item/storage/fancy/candle_box(src)
 	return
 
-/obj/structure/closet/coffin
```
Comment (optimumtact) [inline code/game/objects/structures/crates_lockers/closets/job_closets.dm]: why did you remove the coffin

File: code/modules/cargo/expressconsole.dm (hunk 2 above)
Comment (optimumtact) [reply code/modules/cargo/expressconsole.dm]: tgui will handle it with the default interaction type iirc.

File: code/game/objects/structures/crates_lockers/closets/job_closets.dm (hunk 4 above)
Comment (MrDoomBringer) [reply code/game/objects/structures/crates_lockers/closets/job_closets.dm]: its a refactor, I mentioned it in the PR

Comment (MrDoomBringer) [reply code/game/objects/structures/crates_lockers/closets/job_closets.dm]: @optimumtact Check crates.dm

File: code/modules/cargo/expressconsole.dm [hunk 5]
```diff
@@ -68,29 +87,57 @@
...
+		if("LZCargo")	
+			usingBeacon = FALSE
+			if (beacon)
+				beacon.update_status(SP_UNREADY) //ready light on beacon will turn off
+		if("LZBeacon")
+			usingBeacon = TRUE
+			if (beacon)
+				beacon.update_status(SP_READY) //turns on the beacon's ready light
+		if("printBeacon")//cooldown conditional is located in cargo_express.ract
```
Comment (AnturK) [inline code/modules/cargo/expressconsole.dm]: So it's spoofable.

File: code/modules/cargo/expressconsole.dm [hunk 6]
```diff
@@ -68,29 +87,57 @@
...
+			cooldown = 10//a ~ten second cooldown for printing beacons to prevent spam
+			var/obj/item/supplypod_beacon/C = new /obj/item/supplypod_beacon(loc)
```
Comment (AnturK) [inline code/modules/cargo/expressconsole.dm]: loc should be drop_location()

File: code/modules/cargo/expressconsole.dm [hunk 7]
```diff
@@ -68,29 +87,57 @@
...
+			C.link_console(src)//rather than in beacon's Initialize(), we can assign the computer to the beacon by reusing this proc)
+			printed_beacons++//printed_beacons starts at 0, so the first one out will be called beacon # 1
+			beacon.name = "Supply Pod Beacon #[printed_beacons]"
+			SSshuttle.points -= BEACON_COST
```
Comment (AnturK) [inline code/modules/cargo/expressconsole.dm]: Shouldn't this check if you actually have these points.

File: code/modules/cargo/supplypod_beacon.dm [hunk 8]
```diff
@@ -0,0 +1,109 @@
...
+	switch(consoleStatus)
+		if (SP_LINKED)
+			linked = TRUE
+			playsound(src,'sound/machines/twobeep.ogg',50,0)
+		if (SP_READY)
+			ready = TRUE
+		if (SP_LAUNCH)
+			launched = TRUE
+			playsound(src,'sound/machines/triple_beep.ogg',50,0)
+			playsound(loc,'sound/machines/warning-buzzer.ogg',50,0)
```
Comment (AnturK) [inline code/modules/cargo/supplypod_beacon.dm]: src

File: code/modules/cargo/supplypod_beacon.dm [hunk 9]
```diff
@@ -0,0 +1,109 @@
...
+			playsound(src,'sound/machines/synth_no.ogg',50,0)
+		if (SP_UNREADY)
+			ready = FALSE
+	update_icon()
+
+/obj/item/supplypod_beacon/update_icon()
+	cut_overlays()
+	if (launched)
+		add_overlay("sp_green")
+		addtimer(CALLBACK(src, .proc/endLaunch), 33)//wait 3.3 seconds (time it takes for supplypod to land), then update icon
```
Comment (AnturK) [inline code/modules/cargo/supplypod_beacon.dm]: Why does updating icon handle state. This is very messy.

File: code/modules/cargo/supplypod_beacon.dm [hunk 10]
```diff
@@ -0,0 +1,109 @@
...
+		if (express_console != O)
+			link_console(O)
+
+/obj/item/supplypod_beacon/AltClick(mob/user)
+	if (!user.canUseTopic(src, !issilicon(user)))
+		return
+	if (express_console)
+		unlink_console()
+	else
+		visible_message("<span class='notice'>There is no linked console!</span>")
```
Comment (AnturK) [inline code/modules/cargo/supplypod_beacon.dm]: Shouldn't this be to_chat

File: code/modules/cargo/supplypod_beacon.dm [hunk 11]
```diff
@@ -0,0 +1,109 @@
...
+
+/obj/item/supplypod_beacon/attackby(obj/item/W, mob/user)
+	if(istype(W, /obj/item/pen)) //give a tag that is visible from the linked express console
+		var/tag = stripped_input(user, "What would you like the tag to be?")
```
Comment (AnturK) [inline code/modules/cargo/supplypod_beacon.dm]: Please don't use tag as variable name.

**Review by AnturK — CHANGES_REQUESTED** [review]
Stuff, also what's the state of this.

File: code/modules/cargo/expressconsole.dm (hunk 5 above)
Comment (MrDoomBringer) [reply code/modules/cargo/expressconsole.dm]: whups, i fixed that earlier and didnt remember to change the comment.

File: code/modules/cargo/expressconsole.dm (hunk 7 above)
Comment (MrDoomBringer) [reply code/modules/cargo/expressconsole.dm]: there is a check for this on line 98, does that still work?

File: code/modules/cargo/supplypod_beacon.dm (hunk 11 above)
Comment (MrDoomBringer) [reply code/modules/cargo/supplypod_beacon.dm]: i stole this code from something else. why not use tag?

**Comment by MrDoomBringer** [comment]
@AnturK state of this shooould be ready to merge! Also issues fixed

**Comment by optimumtact** [comment]
\>conflicts
\>ready to merge

File: code/modules/cargo/supplypod_beacon.dm (hunk 11 above)
Comment (AnturK) [reply code/modules/cargo/supplypod_beacon.dm]: It's very easy to mistake/conflict it with the inbuilt tag variable.

**Comment by MrDoomBringer** [comment]
LISTEN THERES A CONFLICT EVERY 10 MINUTES REE @optimumtact

File: code/modules/cargo/supplypod_beacon.dm [hunk 12]
```diff
@@ -0,0 +1,109 @@
...
+		express_console.beacon = null
+	return ..()
+
+/obj/item/supplypod_beacon/proc/unlink_console()
+	if(express_console)
+		express_console.beacon = null
+		express_console = null
+	update_status(SP_UNLINK)
+	update_status(SP_UNREADY) 
+	visible_message("<span class='notice'>[name] has been unlinked from [express_console].</span>")
```
Comment (SpaceManiac) [inline code/modules/cargo/supplypod_beacon.dm]: Writing `[name]` is usually bad practice, should be `[src]`. The capitalized name makes it a proper noun so it'll still come out right.

Also this is run after `express_console` is set to null, so the message is wrong.

File: code/modules/cargo/supplypod_beacon.dm [hunk 13]
```diff
@@ -0,0 +1,109 @@
...
+	if (express_console)//if this beacon has an express console
+		express_console.beacon = null//remove the connection the expressconsole has from beacons
+	express_console = C//set the linked console var to the console
+	express_console.beacon = src//out with the old in with the news
+	update_status(SP_LINKED)
+	if (express_console.usingBeacon)
+		update_status(SP_READY)
+	visible_message("<span class='notice'>[name] linked to [C].</span>")
+
+/obj/item/supplypod_beacon/afterattack(obj/O, mob/user, proximity)
```
Comment (SpaceManiac) [inline code/modules/cargo/supplypod_beacon.dm]: This needs to be changed so that linking a beacon doesn't hit the console with it. I recommend moving it to `/obj/machinery/computer/cargo/express/attackby`.

File: tgui/src/interfaces/cargo_express.ract [hunk 14]
```diff
@@ -21,6 +21,13 @@
 
 {{#if !data.locked }}
   <ui-display title='Express Cargo Console'>
+
+    <ui-section label='Landing Location:'>
```
Comment (SpaceManiac) [inline tgui/src/interfaces/cargo_express.ract]: Remove colon, this comes out "Landing Location::".

**Comment by optimumtact** [comment]
where are you at with this?

File: code/game/objects/structures/crates_lockers/closets/secure/supplypod.dm [hunk 15]
```diff
@@ -1,5 +1,6 @@
 //The "BDPtarget" temp visual is created by the expressconsole, which in turn makes two things: a falling droppod animation, and the droppod itself.
-
+#define POD_STANDARD 0
+#define POD_BLUESPACE 1
```
Comment (SpaceManiac) [inline code/game/objects/structures/crates_lockers/closets/secure/supplypod.dm]: You `#define` these three times each, pick one location for them and define them there.

**Review by SpaceManiac — CHANGES_REQUESTED** [review]
In the future I would suggest atomizing a little more if you want to reduce conflicts and make things easier on maintainers, the supply pod beacons and the religion/coffin changes could easily be two separate PRs.

File: code/modules/cargo/supplypod_beacon.dm [hunk 16]
```diff
@@ -0,0 +1,109 @@
...
+	if (C.beacon)//if new console has a beacon, then...
+		C.beacon.unlink_console()//unlink the old beacon from new console
+	if (express_console)//if this beacon has an express console
+		express_console.beacon = null//remove the connection the expressconsole has from beacons
+	express_console = C//set the linked console var to the console
+	express_console.beacon = src//out with the old in with the news
+	update_status(SP_LINKED)
+	if (express_console.usingBeacon)
+		update_status(SP_READY)
+	visible_message("<span class='notice'>[name] linked to [C].</span>")
```
Comment (SpaceManiac) [inline code/modules/cargo/supplypod_beacon.dm]: The `visible_message` is invisible when actually used due to the item being in a mob's hand. Should probably be a `to_chat` to the person doing the linking. Same `[name]` thing also.

File: code/modules/cargo/supplypod_beacon.dm (hunk 12 above)
Comment (SpaceManiac) [reply code/modules/cargo/supplypod_beacon.dm]: Actually seems odd that this is a `visible_message` at all. There's nobody to reasonably `to_chat` it to, but the icon already goes dark, you could reasonably let this be silent.

**Review by SpaceManiac — CHANGES_REQUESTED** [review]

File: code/modules/cargo/supplypod_beacon.dm (hunk 13 above)
Comment (MrDoomBringer) [reply code/modules/cargo/supplypod_beacon.dm]: oh thanks! i was actually just wondering how to fix that

File: code/modules/cargo/supplypod_beacon.dm (hunk 16 above)
Comment (MrDoomBringer) [reply code/modules/cargo/supplypod_beacon.dm]: i can pass `usr` from tgui into a mob/living/user param right?

Comment (SpaceManiac) [reply code/modules/cargo/supplypod_beacon.dm]: `usr` is guaranteed to be a mob but it might be a ghost (`/mob/dead`) if an admin is using AI interaction. If all you're doing is `to_chat` it's fine.

File: tgui/src/interfaces/cargo_express.ract (hunk 14 above)
Comment (MrDoomBringer) [reply tgui/src/interfaces/cargo_express.ract]: how the fuk did i miss this

**Comment by MrDoomBringer** [comment]
alrighty oranges i had to drop a light rebase here or there and meet some requested changes but i just tested it and it seems to work as intended. guess now we see if spacemaniac approves

File: code/modules/cargo/expressconsole.dm [hunk 17]
```diff
@@ -1,41 +1,63 @@
...
-		podID = 1//doesnt effect circuit board, so that reversal is possible
+		podID = POD_BLUESPACE//doesnt effect circuit board, making reversal possible
 		to_chat(user, "<span class='notice'>You insert the disk into [src], allowing for advanced supply delivery vehicles.</span>")
 		qdel(W)
 		return TRUE
+	else if(istype(W, /obj/item/supplypod_beacon))
+		var/obj/item/supplypod_beacon/sb = W
+		if (sb.express_console != src)
+			sb.link_console(src, user)
+			return TRUE
```
Comment (SpaceManiac) [inline code/modules/cargo/expressconsole.dm]: Should move this out of the `if` so already being linked doesn't make you hit the console.

File: code/modules/cargo/expressconsole.dm [hunk 18]
```diff
@@ -108,25 +158,28 @@
...
+						if (istype(beacon) && usingBeacon)
+							LZ = get_turf(beacon)
+							beacon.update_status(SP_LAUNCH)
 						SSshuttle.points -= SO.pack.cost
+
 						new /obj/effect/DPtarget(LZ, SO, podID)
 						. = TRUE
 						update_icon()
 			else
 				if(SO.pack.cost * (0.72*MAX_EMAG_ROCKETS) <= SSshuttle.points) // bulk discount :^)
```
Comment (SpaceManiac) [inline code/modules/cargo/expressconsole.dm]: The number on this line is different than the one on line 189, is this intentional?

File: code/modules/cargo/expressconsole.dm [hunk 19]
```diff
@@ -108,25 +158,28 @@
 				rank = "Silicon"
 			var/reason = ""
 			var/list/empty_turfs
-			var/area/landingzone
 			var/datum/supply_order/SO = new(pack, name, rank, ckey, reason)
 			if(!(obj_flags & EMAGGED))
 				if(SO.pack.cost <= SSshuttle.points)
 					landingzone = locate(/area/quartermaster/storage) in GLOB.sortedAreas
```
Comment (SpaceManiac) [inline code/modules/cargo/expressconsole.dm]: Runtimes if the map doesn't have this area on it, preventing any use of the console.

Fall back to the area in which the console is located instead?

File: code/modules/cargo/expressconsole.dm [hunk 20]
```diff
@@ -108,25 +158,28 @@
...
-					for(var/turf/open/floor/T in landingzone.contents)
+					for(var/turf/open/floor/T in landingzone.contents)//uses default landing zone
 						if(is_blocked_turf(T))
 							continue
 						LAZYADD(empty_turfs, T)
 						CHECK_TICK
 					if(empty_turfs && empty_turfs.len)
 						var/LZ = empty_turfs[rand(empty_turfs.len-1)]
+						if (istype(beacon) && usingBeacon)
+							LZ = get_turf(beacon)
```
Comment (SpaceManiac) [inline code/modules/cargo/expressconsole.dm]: Won't get reached, and therefore the beacon won't be usable, if the cargo area is entirely blocked.

**Review by SpaceManiac — CHANGES_REQUESTED** [review]

File: code/modules/cargo/expressconsole.dm (hunk 17 above)
Comment (MrDoomBringer) [reply code/modules/cargo/expressconsole.dm]: that was intentional: players would recieve a response when unsuccessfully linking the device

File: code/modules/cargo/expressconsole.dm (hunk 18 above)
Comment (MrDoomBringer) [reply code/modules/cargo/expressconsole.dm]: o shit

File: code/modules/cargo/expressconsole.dm (hunk 17 above)
Comment (SpaceManiac) [reply code/modules/cargo/expressconsole.dm]: It's fine to get a chat message but a failed link shouldn't also attack the console.

File: code/modules/cargo/expressconsole.dm (hunk 19 above)
Comment (SpaceManiac) [reply code/modules/cargo/expressconsole.dm]: If the intended area doesn't exist then falling back to the current area and logging a warning is probably sane. Look around for other code which warns.

File: code/modules/cargo/expressconsole.dm [hunk 21]
```diff
@@ -162,21 +162,26 @@
...
 						CHECK_TICK
+					var/LZ
 					if(empty_turfs && empty_turfs.len)
-						var/LZ = empty_turfs[rand(empty_turfs.len-1)]
-						if (istype(beacon) && usingBeacon)
-							LZ = get_turf(beacon)
-							beacon.update_status(SP_LAUNCH)
-						SSshuttle.points -= SO.pack.cost
+						LZ = empty_turfs[rand(empty_turfs.len-1)]
+					else if (istype(beacon) && usingBeacon)
```
Comment (SpaceManiac) [inline code/modules/cargo/expressconsole.dm]: This probably needs to be moved up above the `empty_turfs` check, as is a usable cargo bay area will prevent use of becons.

**Review by SpaceManiac — CHANGES_REQUESTED** [review]

File: code/modules/cargo/expressconsole.dm (hunk 21 above)
Comment (MrDoomBringer) [reply code/modules/cargo/expressconsole.dm]: goddamn im dumb thanks

**Review by SpaceManiac — CHANGES_REQUESTED** [review]
One minor point upon a coderead but otherwise looks good.

File: code/modules/cargo/expressconsole.dm [hunk 22]
```diff
@@ -68,29 +90,57 @@
...
+			usingBeacon = TRUE
+			if (beacon)
+				beacon.update_status(SP_READY) //turns on the beacon's ready light
+		if("printBeacon")
+			cooldown = 10//a ~ten second cooldown for printing beacons to prevent spam
+			var/obj/item/supplypod_beacon/C = new /obj/item/supplypod_beacon(drop_location())
+			C.link_console(src, usr)//rather than in beacon's Initialize(), we can assign the computer to the beacon by reusing this proc)
+			printed_beacons++//printed_beacons starts at 0, so the first one out will be called beacon # 1
+			beacon.name = "Supply Pod Beacon #[printed_beacons]"
+			SSshuttle.points -= BEACON_COST//check for this is under ui_data
```
Comment (SpaceManiac) [inline code/modules/cargo/expressconsole.dm]: UI checks can't be fully trusted, someone might spamclick this button multiple times before the UI updates. You still need a conditional here.

File: code/modules/cargo/expressconsole.dm [hunk 23]
```diff
@@ -108,32 +158,41 @@
...
 					landingzone = locate(/area/quartermaster/storage) in GLOB.sortedAreas
-					for(var/turf/open/floor/T in landingzone.contents)
+					if (!landingzone)
+						WARNING("[src] couldnt find a Quartermaster/Storage area on the station, and as such has set the LZ to the area it resides in.")
+						landingzone = get_area(src)
+					for(var/turf/open/floor/T in landingzone.contents)//uses default landing zone
 						if(is_blocked_turf(T))
 							continue
 						LAZYADD(empty_turfs, T)
 						CHECK_TICK
```
Comment (SpaceManiac) [inline code/modules/cargo/expressconsole.dm]: This `CHECK_TICK` means it might be possible for two invocations of this function to overlap; you need an additional check at the exact moment you remove credits (L182), otherwise spam-clicking can cause credits to go negative.

**Review by SpaceManiac — APPROVED** [review]

File: config/admins.txt [hunk 24]
```diff
@@ -136,3 +136,4 @@ Averagejoe82 = Game Master
 The Dreamweaver = Game Master
 88Naoki = Game Master
 Naksuasdf = Game Master
+MrDoomBringer = Game Master
```
Comment (Qustinnus) [inline config/admins.txt]: wtf mrdoombringer trying to take over tg??

Comment (MrDoomBringer) [reply config/admins.txt]: delete this comment right now PLEASE ILL DO ANYTHING
